test/functional/mining.py

```diff
@@ -0,0 +1,336 @@
+                for x in new_template['transactions']:
+                    # Note that we're using the original mempool modified fees,
+                    # not the current ones!
+                    new_fee += int(mempool[x['txid']]['modifiedfee']*COIN)
+                self.log.debug("Recent transactions: block_reward=%d non-recent=%d all=%d", block_reward, new_fee, block_fee)
+
+                assert((1-recent_tx_stale_rate)*(block_reward+block_fee) > (block_reward+new_fee))
+                # Re-prioritise everything
+                for txid, entry in mempool.items():
+                    if (entry["time"] > mock_time - recent_tx_window):
+                        node.prioritisetransaction(txid, int(entry["fee"]*COIN))
+                        if (txid in template_txids):
+                            ancestors = node.getmempoolancestors(txid)
+                            for a in ancestors:
+                                if (mempool[a]['time'] < mock_time - recent_tx_window):
+                                    node.prioritisetransaction(a, int(-(entry["fee"]*COIN)))
+
+
+            mock_time -= recent_tx_window
+
+        if (blocks_with_recent_tx == 0 or blocks_without_recent_tx == 0):
+            self.log.info("Warning - test_recent_transactions_excluded(): only one branch covered (blocks_with_recent_tx = %d, blocks_without_recent_tx = %d)", blocks_with_recent_tx, blocks_without_recent_tx)
+
+        # Get rid of the mocktime
+        node.setmocktime(0)
+
+    def run_test(self):
+        # Leave IBD and generate some coins to spend.
+        # Give everyone plenty of coins
+        self.log.info("Generating initial coins for all nodes")
+        for i in range(2):
+            for x in self.nodes:
+                # Building a long blockchain here will make the block reward
+                # smaller, compared to the fees -- this is helpful for
+                # test_recent_transactions_excluded()
+                x.generate(201)
+                self.sync_all()
+
+        # Add some segwit coins to everyone
+        self.log.info("Generating segwit coins for all nodes")
+        self.create_segwit_coins()
+
+        # Run tests...
+        self.log.info("Populating mempool with a lot of transactions")
+        self.populate_mempool(500)
+
+        self.log.info("Running test_max_block_weight")
+        self.test_max_block_weight(self.nodes[0])
+
+        self.sync_all()
+        self.log.info("Running test_recent_transactions_excluded")
+        for i in range(3):
+            self.test_recent_transactions_excluded(self.nodes[i], recent_tx_window=self.cnb_args[i]['window'], recent_tx_stale_rate=self.cnb_args[i]['stalerate'])
+
+        self.sync_all()
+        self.log.info("Running test_ancestor_feerate_sort")
+        self.test_ancestor_feerate_sort(self.nodes[0], recent_tx_window=self.cnb_args[i]['window'])
```

Comment:
In commit "Add functional test for CreateNewBlock"

`i` should be `0` here I think